### PR TITLE
Possibility to subscribe for proposal changes

### PR DESCRIPTION
**Pull request by Waldz**
Possibility for mobile to update proposal selector wo explicit fetch

### CODE REVIEW

File: core/discovery/event.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package discovery
+
+// Topic represents the different topics a consumer can subscribe to
+const (
+	// ProposalAddedEventTopic represents newly announced proposal
+	ProposalAddedEventTopic = "ProposalAdded"
```
Comment (anjmao) [inline core/discovery/event.go]: Go's naming convention for multiple constants is to to prefix. See example https://github.com/golang/go/blob/c04e47f82159f1010d7403276b3dff5ab836fd00/src/net/http/status.go#L9

```go
const (
    EventTopicProposalAdded = "ProposalAdded"
    EventTopicProposalUpdated = "ProposalUpdated"
    EventTopicProposalRemovedE = "ProposalRemoved"
)
```

When it is more easy to find all topics when writing in IDE `discovery.EventTopic*`

File: core/discovery/storage.go [hunk 2]
```diff
@@ -41,14 +41,17 @@ type ProposalFilter interface {
 }
 
 // NewStorage creates new instance of ProposalStorage
-func NewStorage() *ProposalStorage {
+func NewStorage(eventPublisher Publisher) *ProposalStorage {
```
Comment (anjmao) [inline core/discovery/storage.go]: I saw Publisher interface in 10 places. There is already pacakge eventbus which contains this interface

```go
package eventbus

// Publisher is responsible for publishing given events.
type Publisher interface {
	Publish(topic string, data interface{})
}
```

I suggest to remove all other interface copies and use this one.

File: mobile/mysterium/entrypoint.go [hunk 3]
```diff
@@ -212,6 +207,35 @@ func (mb *MobileNode) GetProposal(req *GetProposalRequest) ([]byte, error) {
...
 
+// ProposalChangeCallback represents proposal callback.
+type ProposalChangeCallback interface {
+	OnChange(proposal []byte)
+}
+
+// RegisterProposalAddedCallback registers callback which is called on newly announced proposals
+func (mb *MobileNode) RegisterProposalAddedCallback(cb ProposalChangeCallback) {
+	_ = mb.eventBus.SubscribeAsync(discovery.ProposalAddedEventTopic, func(proposal market.ServiceProposal) {
+		proposalPayload, _ := mb.proposalsManager.mapToProposalResponse(&proposal)
```
Comment (anjmao) [inline mobile/mysterium/entrypoint.go]: Would be good to log possible error.

File: mobile/mysterium/entrypoint.go [hunk 4]
```diff
@@ -212,6 +207,35 @@ func (mb *MobileNode) GetProposal(req *GetProposalRequest) ([]byte, error) {
...
+		cb.OnChange(proposalPayload)
```
Comment (anjmao) [inline mobile/mysterium/entrypoint.go]: Sad news are that mapToProposalResponse is not adding quality level data. Before it was used only for single GetProposal which doesn't need quality data. But now these proposals need to have such data as they will update actual UI. See `func (m *proposalsManager) addQualityData(proposals []*proposal)` for usage how it was done. Probably quality metrics need to be cached on initial GetProposals etc. to be accessible for each event.

File: mobile/mysterium/entrypoint.go [hunk 5]
```diff
@@ -212,6 +207,35 @@ func (mb *MobileNode) GetProposal(req *GetProposalRequest) ([]byte, error) {
...
+
+// RegisterProposalAddedCallback registers callback which is called on re-announced proposals
+func (mb *MobileNode) RegisterProposalUpdatedCallback(cb ProposalChangeCallback) {
+	_ = mb.eventBus.SubscribeAsync(discovery.ProposalUpdatedEventTopic, func(proposal market.ServiceProposal) {
+		proposalPayload, _ := mb.proposalsManager.mapToProposalResponse(&proposal)
+		cb.OnChange(proposalPayload)
+	})
+}
+
+// RegisterProposalAddedCallback registers callback which is called on re-announced proposals
```
Comment (zolia) [inline mobile/mysterium/entrypoint.go]:
```suggestion
// RegisterProposalRemovedCallback registers callback which is called on unregistered proposals
```

File: core/discovery/event.go (hunk 1 above)
Comment (Waldz) [reply core/discovery/event.go]: Named by existing event convention. Renamed

File: core/discovery/storage.go (hunk 2 above)
Comment (Waldz) [reply core/discovery/storage.go]: Improved, just not sure what it improves. Usually Go encourages to have interfaces on caller side.

File: core/discovery/storage.go [hunk 6]
```diff
@@ -85,6 +88,19 @@ func (s *ProposalStorage) Set(proposals []market.ServiceProposal) {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
+	proposalsOld := s.proposals
```
[the remark this answers is not in the document]
Comment (Waldz) [reply core/discovery/storage.go]: Simplified parallelism

File: mobile/mysterium/entrypoint.go (hunk 4 above)
Comment (Waldz) [reply mobile/mysterium/entrypoint.go]: Looks like need to addressed in another PR

Comment (anjmao) [reply mobile/mysterium/entrypoint.go]: Could be possible addressed in https://github.com/mysteriumnetwork/node/issues/1524

**Pull request merged**